system/lib/compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_emscripten.cc

```diff
@@ -0,0 +1,63 @@
+//===-- sanitizer_stacktrace_emscripten.cc --------------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file is shared between AddressSanitizer and ThreadSanitizer
+// run-time libraries.
+//
+// Implementation of fast stack unwinding for Emscripten.
+//===----------------------------------------------------------------------===//
+
+#ifdef __EMSCRIPTEN__
+
+#include "sanitizer_common.h"
+#include "sanitizer_stacktrace.h"
+
+namespace __sanitizer {
+
+extern "C" {
+  void emscripten_stack_snapshot();
+  uptr emscripten_stack_unwind(uptr pc, int level);
+  uptr emscripten_return_address(int level);
+}
+
+uptr StackTrace::GetCurrentPc() {
+  emscripten_stack_snapshot();
+  return GET_CALLER_PC();
+}
+
+void BufferedStackTrace::FastUnwindStack(uptr pc, uptr bp, uptr stack_top,
+                                         uptr stack_bottom, u32 max_depth) {
+  bool saw_pc = false;
+  trace_buffer[0] = pc;
+  size = 1;
+
+  uptr pc1;
+  if (emscripten_stack_unwind(pc, 0)) {
+    // If we have the stack trace cached with this pc.
```

Comment:
Can't the same PC appear in multiple stack traces? I don't understand why they are being cached.

Reply:
The comment is misleading.

`emscripten_stack_unwind` will only unwind the stack if the `pc` passed matches the return address of the last call to `emscripten_stack_snapshot`.

This is so that it unwinds from the cached stack trace in `UNWIND_CACHE.last_stack` instead of the current stack, which is necessary for when the caller uses `StackTrace::GetCurrentPc()` instead of `__builtin_return_address`. In this case, when the stack tracer is called, the PC in the caller has changed, and taking the stack trace in the tracer will not find the `pc` passed.